example/example/spiders/books_06.py

```diff
@@ -0,0 +1,75 @@
+# -*- coding: utf-8 -*-
+"""
+Scrapy spider with reusable components used for two types of pages:
+* Listings page, with the components: book list and pagination
+* Detail book page, with the components: the item itself, the breadcrumbs and
+  the recently viewed list of books
+
+All links to detail book pages are found are followed.
+All links to search result pages are followed as well.
+"""
+
+import scrapy
+import attr
+
+from scrapy_po import WebPage, ItemPage
+
+
+class ListingsPiece(WebPage):
+    def urls(self):
+        return self.css('.image_container a::attr(href)').getall()
+
+
+class PaginationPiece(WebPage):
+    def urls(self):
+        return self.css('.pager a::attr(href)').getall()
+
+
+class BreadcrumbsPiece(WebPage):
+    def urls(self):
+        return self.css('.breadcrumb a::attr(href)').getall()
+
+
+@attr.s(auto_attribs=True)
+class ListingsPage(WebPage):
+    book_list: ListingsPiece
+    pagination: PaginationPiece
+
+
+@attr.s(auto_attribs=True)
+class BookPage(WebPage, ItemPage):
+    recently_viewed: ListingsPiece
+    breadcrumbs: BreadcrumbsPiece
+
+    def to_item(self):
+        return {
+            'url': self.url,
+            'name': self.css("title::text").get(),
+        }
+
+
+class BooksSpider(scrapy.Spider):
+    name = 'books_06'
+    start_urls = ['http://books.toscrape.com/']
+
+    def parse(self, response, page: ListingsPage):
+        """ Callback for Listings pages """
+        yield from self.follow_listing(response, page.book_list)
+        yield from self.follow_pagination(response, page.pagination)
```

Comment:
Extracting this code in a spider to separate methods makes sense. I wonder if we can have Scrapy-specific wrappers around Paginator&friends, which provide a method to get scrapy.Requests instead of just urls. Code could look like this then:

```py
def parse(self, response, page: ListingsPage):
    yield from page.book_list.requests(self.parse_book)
    yield from page.pagination.requests(self.parse, priority=+10)
```

Wrapping can be either implicit (wrap everythig which callbacks receive) or, better, explicit; a few options:

```py
class ScrapyListingPage(ListingsPage, RequestsFromUrlsMixin):
    pass

ScrapyListingPage = scrapy_po.add_requests_method(ListingsPage)
```

or even just this (requires upcoming scrapy 2.0 - see https://github.com/scrapy/scrapy/pull/4057):
```py
yield from response.follow_all(page.pagination.urls(), self.parse_book)
```

It is better to handle this separately though, not in this PR, unless we switch to scrapy 2.0.

Reply:
I see. I understand. Makes sense. I'm going to create a PR with the explicit variant you proposed.

> unless we switch to scrapy 2.0.

I have not enough confidence/knowledge to say if we should or not. Should we?

Reply:
Filled a PR here: https://github.com/scrapinghub/scrapy-po/pull/9

Reply:
I think we should target Scrapy 2.0 in examples, but provide compatibility with Scrapy 1.8 if possible (it doesn't have to have all the same usability features - i.e. no need to backport response.follow_all). It is going to be released very soon, maybe even this week.

Scrapy 2.0 is good, because it should allow to use https://github.com/scrapinghub/scrapinghub-autoextract asyncio interface for Scrapy integration, which provides a proper retry feature / handling of rate limits.

For examples it shouldn't matter that they're not compatible with Scrapy < 2.0.

Reply:
See https://github.com/scrapy/scrapy/blob/034e2c31c7d55333c3de208f80dcee1bf45ef9b9/scrapy/http/response/text.py#L122

Reply:
Sorry, I should have thought about it more. I think switching to Scrapy 2.0 in examples is better than adding custom shortcuts to the codebase.

Reply:
As you commented in a different PR, better to just migrate the examples to Scrapy 2.0. But let's do it better in a separate PR.